content/blog/2017/2017-02-23-declarative-saucelabs-xunit.adoc

```diff
@@ -0,0 +1,399 @@
+This would run the literal command string shown and the values would be substituted by the shell.
+
+
+== Moving Complex Code to Shared Libraries
+
+Now that we have settings separated from the code, we can do some code clean up.
+Unlike the previous post, we don't have any repeating code,
+but we do have some distractions.
+The `sauce`, `sauceconnect`, and `nightwatch` steps are only used in unison in this project,
+and that xUnit `step` is a bit ugly as well.
+Let's move those into our shared library as custom steps with parameters.
+We'll change the `Jenkinsfile` in our main project,
+and add the custom steps to a branch named
+`blog/declarative/sauce` in our library repository.
+
+.Jenkinsfile
+[source, groovy]
+----
+@Library('bitwiseman-shared@blog/declarative/sauce') _
+
+/* ... unchanged ... */
+
+stage ("Test") {
+    steps {
+        sauceNightwatch sauceCredential: saucelabsCredentialId,
+            platforms: ${platformConfigs},
+            testFilter: ${sauceTestFilter}
+        }
+    }
+    post {
+        always {
+            xUnitPublishResults pattern: 'reports/**',
+                failWhenSkippedExceeds: 0,
+                failWhenFailedExceeds: 10
+
+            saucePublisher()
+        }
+    }
+}
+----
+
+.vars/sauceNightwatch.groovy
+[source, groovy]
+----
+def call(String sauceCredential, String platforms = null, String testFilter = null) {
+    platforms = platforms ? "-e '" + platforms + "'" : '';
+    testFilter = testFilter ? "--test '" + testFilter + "'" : '';
+
+    // Add sauce credentials
+    sauce(sauceCredential) {
+        // Start sauce connect
+        sauceconnect() {
+            // Run selenium tests using Nightwatch.js
+            // Ignore error codes. The junit publisher will cover setting build status.
+            sh "./node_modules/.bin/nightwatch ${platforms} ${testFilter} || true" // <1>
+        }
+    }
+}
+----
+<1> In this form, this could not be written using a literal single-quoted string.
+Here, `platforms` and `testFilter` are groovy variables, not environment variables.
+
+.vars/xUnitPublishResults.groovy
+[source, groovy]
+----
+def call(String pattern, Integer failWhenSkippedExceeds,
+        Integer failWhenFailedExceeds) {
+    step([$class: 'XUnitBuilder',
+        thresholds: [
+            [$class: 'SkippedThreshold', failureThreshold: failWhenSkippedExceeds],
+            // Allow for a significant number of failures
+            // Keeping this threshold so that overwhelming failures are guaranteed
+            //     to still fail the build
+            [$class: 'FailedThreshold', failureThreshold: failWhenFailedExceeds]],
+        tools: [[$class: 'JUnitType', pattern: pattern]]])
+}
+----
+
+
+== Running Conditional Stages with `when`
+
+This is a sample web testing project.
+We probably wouldn't deploy it like we would production code,
+but we might still want to deploy or publish it, to a artifact repository for example.
```

Comment:
*an* artifact repository

Perhaps a command missing before 'for example', unsure.